cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go

```diff
@@ -119,9 +133,21 @@ func (as *AgentPool) MaxSize() int {
 
 // GetVMIndexes gets indexes of all virtual machines belonging to the agent pool.
 func (as *AgentPool) GetVMIndexes() ([]int, map[int]string, error) {
-	instances, err := as.GetVirtualMachines()
-	if err != nil {
-		return nil, nil, err
+	virtualMachinesStatusCache.mutex.Lock()
+	defer virtualMachinesStatusCache.mutex.Unlock()
+
+	var instances []compute.VirtualMachine
+	var err error
+
+	if virtualMachinesStatusCache.lastRefresh.Add(vmInstancesRefreshPeriod).After(time.Now()) {
```

Comment:
there're three places of such codes, let's extract a new func for getting VMs from cache and refresh if the cache is outdated.

Reply:
Finished.